src/test/java/com/aws/greengrass/authorization/WildcardVariableTrieTest.java

```diff
@@ -0,0 +1,187 @@
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@ExtendWith(GGExtension.class)
+class WildcardVariableTrieTest {
+    @Test
+    void testGlobWildcardMatching() {
+        WildcardVariableTrie rt = new WildcardVariableTrie();
+        rt.add("abc*xyz*");
+        rt.add("*/def");
+        assertTrue(rt.matches("abc1234asdxyz456", true));
+        assertTrue(rt.matches("abcdxyz", true));
+        assertTrue(rt.matches("/def", true));
+        assertTrue(rt.matches("12345/def", true));
+        assertTrue(rt.matches(null, true));
+        assertFalse(rt.matches("abc123xyz456/89", true));
+        assertFalse(rt.matches("2/3/def", true));
+        assertFalse(rt.matches("", true));
+
+        // Only '*' should match all resources
+        WildcardVariableTrie rt1 = new WildcardVariableTrie();
+        rt1.add("*");
+        rt1.add("*wer");
+        assertTrue(rt1.matches("9999/88", true));
+
+        WildcardVariableTrie rt2 = new WildcardVariableTrie();
+        rt2.add("a/*/*/d");
+        rt2.add("**/abc");
+        assertTrue(rt2.matches("a/12/34/d", true));
+        assertFalse(rt2.matches("a/2/3/4/d", true));
+        assertTrue(rt2.matches("78/abc", true));
+        assertFalse(rt2.matches("7/8/abc", true));
+    }
+
+    @Test
+    void testMQTTMultilevelWildcardMatching() {
+        //test MQTT wildcard usages according to MQTT V5.0 spec (https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html)
+        // Test Valid usages
+        WildcardVariableTrie rt = new WildcardVariableTrie();
+        rt.add("abc/#");
+        rt.add("123/#/xyz/#");
+        rt.add("#/zzz/45");
+        assertTrue(rt.matches("abc/1/2/3", true));
+        assertTrue(rt.matches("abc", true));
+        assertTrue(rt.matches("abc/", true));
+        assertTrue(rt.matches("abc/4/5/6", true));
+        assertTrue(rt.matches("123/#/xyz", true));
+        assertTrue(rt.matches("123/#/xyz/1/2/345", true));
+        assertFalse(rt.matches("abcd/e/f123/4/5/6", true));
+        assertFalse(rt.matches("123/4/5/6/xyz/abc", true));
+        assertFalse(rt.matches("12/34/zzz/45", true));
+        assertFalse(rt.matches("/zzz/45", true));
+
+        // Only '#' should match all resources
+        WildcardVariableTrie rt1 = new WildcardVariableTrie();
+        rt1.add("#");
+        assertTrue(rt1.matches("asdu76/asdas/23", true));
+        assertTrue(rt1.matches("**90io", true));
+
+        WildcardVariableTrie rt2 = new WildcardVariableTrie();
+        rt2.add("/#");
+        assertFalse(rt2.matches("a/b/c/d", true));
+        assertFalse(rt2.matches("asd", true));
+        assertTrue(rt2.matches("", true));
+
+        // Test Invalid usages
+        WildcardVariableTrie rt3 = new WildcardVariableTrie();
+        rt3.add("w/qqq#");
+        rt3.add("12/#/");
+        rt3.add("##");
```

Comment:
retest these without the MQTT wildcard option enabled; verify that these are allowed as literal strings.